frontend/src/app/sessions/user-sessions-wrapper/create-session/create-readonly-session/create-readonly-session.component.ts

```diff
@@ -60,33 +62,82 @@ export class CreateReadonlySessionComponent implements OnInit {
   ) {}
 
   ngOnInit(): void {
-    this.modelService.models$
-      .pipe(untilDestroyed(this), filter(Boolean))
-      .subscribe((models) => (this.models = models));
-
     this.projectService.project$
       .pipe(untilDestroyed(this), filter(Boolean))
       .subscribe((project) => (this.projectSlug = project.slug));
 
+    this.loadToolsAndModels().subscribe(([models, allVersions]) => {
+      this.resolveVersionCompatibility(models, allVersions);
+    });
+
     this.userSessionService.loadSessions();
-    this.toolService.getTools().subscribe();
+  }
+
+  loadToolsAndModels(): Observable<[Model[], ToolVersionWithTool[]]> {
+    return combineLatest([
+      this.modelService.models$.pipe(untilDestroyed(this), filter(Boolean)),
+      this.toolService.getVersionsForTools(),
+    ]).pipe(
+      tap(([_, versions]) => {
+        this.allToolVersions = versions;
+      }),
+    );
+  }
+
+  resolveVersionCompatibility(
+    models: Model[],
+    allVersions: ToolVersionWithTool[],
+  ): void {
+    this.models = [];
+
+    const modelsWithVersion = models.filter((m) => m.version);
+    for (const model of modelsWithVersion) {
+      const extendedModel = model as ModelWithCompatibility;
+      extendedModel.compatibleVersions = [
+        this.findVersionByID(model.version!.id, allVersions)!,
+      ];
+
+      for (const version of allVersions!) {
+        if (version.config.compatible_versions.includes(model.version!.id)) {
+          extendedModel.compatibleVersions.push(version);
+        }
+      }
+
+      this.models?.push(extendedModel);
+    }
+  }
+
+  findVersionByID(id: number, allVersions: ToolVersionWithTool[]) {
+    return allVersions.find((v) => v.id === id);
   }
 
   get tools(): Tool[] | undefined {
-    if (this.models === undefined || this.toolService.tools === undefined) {
+    if (this.models === undefined) {
       return undefined;
     }
-    const toolIds = this.models.map((m) => m.tool.id);
-    return this.toolService.tools.filter((t) => toolIds?.includes(t.id));
+    return this.models
+      ?.map((m) => m.compatibleVersions.map((version) => version.tool))
+      .flat()
+      .filter(
+        (value, index, self) =>
+          index === self.findIndex((tool) => tool.id === value.id),
+      );
   }
 
   requestReadonlySession(): void {
     if (this.toolSelectionForm.valid) {
       const dialogRef = this.dialog.open(CreateReadonlySessionDialogComponent, {
         data: {
           projectSlug: this.projectSlug,
-          models: this.models,
-          modelVersionId: this.toolSelectionForm.value.version!.id,
+          models: this.models?.filter(
+            (model) =>
+              model.compatibleVersions.findIndex(
```

Comment:
Similar to above, it is somewhat confusing and not intuitive to me why we use indices here. So please consider making this clearer in some way.

Reply:
Also outsourced to a function `removeVersionDuplicates`.

Reply:
Outsourced to function `isCompatibleWithSelectedVersion`.